rpki-validator/src/main/java/net/ripe/rpki/validator3/util/HappyEyeballsResolver.java

```diff
@@ -273,52 +270,55 @@ private static Runnable dnsLookupRunnable(Name hostname, int queryType, Queue<Op
-                        final boolean haveMoreAF2 = openConnectionFromQueue(secondAfAddresses);
-                        // watch out! do not inline these booleans,
-                        // otherwise you'll shortcircuit second evaluation
-                        keepGoing = haveMoreAF1 || haveMoreAF2;
-                    }
-                } catch (InterruptedException e) {
-                    // just stop
-                } finally {
-                    socketChannels.add(Optional.empty());
-                }
-            }
+    private static Optional<SocketAddress> findBestAddress(final Queue<Optional<InetAddress>> firstAfAddresses,
+                                                           final Queue<Optional<InetAddress>> secondAfAddresses,
+                                                           final int port, final Selector selector) {
+        final long startTime = System.nanoTime();
+        final long deadline = startTime + HAPPY_EYEBALLS_TOTAL_TIMEOUT_NANOS;
+        Optional<SocketAddress> connectedSocket;
+        boolean keepGoing = true;
+        while (keepGoing && !Thread.currentThread().isInterrupted()) {
+            final boolean haveMoreAF1 = openConnectionFromQueue(firstAfAddresses, port, selector);
+            connectedSocket = awaitSuccessfulConnection(selector, CONNECTION_ATTEMPT_DELAY_MILLIS);
+            if (connectedSocket.isPresent()) return connectedSocket;
 
-            private boolean openConnectionFromQueue(Queue<Optional<InetAddress>> queue) throws InterruptedException {
-                final Optional<InetAddress> addressOptional = queue.poll();
-                if (addressOptional != null) {
-                    if (addressOptional.isPresent()) {
-                        final InetAddress address = addressOptional.get();
-                        try {
-                            log.debug("Connecting to {}", address);
-                            final SocketChannel socketChannel = SocketChannel.open();
-                            socketChannel.configureBlocking(false);
-                            socketChannel.connect(new InetSocketAddress(address, port));
-                            socketChannels.add(Optional.of(socketChannel));
-                        } catch (IOException e) {
-                            // can't connect - move on
-                            log.debug("Error connecting to {}", address);
-                        } finally {
-                            TimeUnit.MILLISECONDS.sleep(CONNECTION_ATTEMPT_DELAY_MS);
-                        }
-                    } else {
-                        return false;
-                    }
+            final boolean haveMoreAF2 = openConnectionFromQueue(secondAfAddresses, port, selector);
+            connectedSocket = awaitSuccessfulConnection(selector, CONNECTION_ATTEMPT_DELAY_MILLIS);
+            if (connectedSocket != null) return connectedSocket;
```

Comment:
Shouldn't this one be the same check for isPresent?

Reply:
Good catch!  This is a bug.